Restyle repository page and checkout instructions

oliverguenther opened the pull request:
Restyles the core checkout functionality according to the visual by Christoph: http://clients.zzmedia.net/openproject/repository/
### Synopsis
- Restyle repositories `show` and `empty` actions
- Allows multiple toolbars to co-exist
- Extends toolbar functionality
  - Including select2 fields (proper scaling). To avoid too small selects, toolbar select2 fields are at least 150px wide
  - Including inline labels in toolbars
  - Prepended and appended inputs
  - Icon-only items (they require font-size changes and a small offset)
- Toggle button in toolbar to show / hide checkout instructions
### TODO
- [X] Folder hierarchy
- [X] Restyle breadcrumbs
- [X] Notifications from clipboard don't propagate correctly. Fixed using `$timeout`.
- [X] ~~Remember user decision when checkout hidden~~ I move this out of this PR and into a separated feature, as its currently blocked by the missing endpoint, which I started specifying in https://github.com/opf/openproject/pull/3457

**Checkout toolbar**
![bildschirmfoto 2015-08-31 um 14 55 01](https://cloud.githubusercontent.com/assets/459462/9579173/c5f5c020-4ff0-11e5-84ab-21db1e9518ac.png)

**Revision toolbar**
![bildschirmfoto 2015-08-31 um 14 55 08](https://cloud.githubusercontent.com/assets/459462/9579174/c7727d3a-4ff0-11e5-907d-3553f467b49c.png)

**Empty repository page**
![bildschirmfoto 2015-09-03 um 10 08 19](https://cloud.githubusercontent.com/assets/459462/9653445/bd43514c-5223-11e5-8f7c-7739f5470bf7.png)

**Notifications**
![bildschirmfoto 2015-09-03 um 10 27 36](https://cloud.githubusercontent.com/assets/459462/9653850/a0e7f7ac-5226-11e5-893e-b98ee728c64b.png)

Relevant work package: https://community.openproject.org/work_packages/21343

oliverguenther committed:
Style checkout instructions and url according to visual

This commit provides:

* The checkout instructions as a notification-box with info icon
* A style for affixed inputs in the toolbar
* The checkout URL as an affixed input
* The checkout URL inside the notifcation-box on empty repository pages
* Animations for the checkout instructions toggle
* Restyle revision section

oliverguenther committed:
Allow multiple toolbars to exist, extend toolbar functionality

Previously, toolbars were thought to be unique on a page and thus
identified by id. As repository now contains multiple, this was changed
to class-based selectors.

This commit also adds functionality in toolbars for:

 * Including select2 fields (proper scaling). To avoid too small
   selects, toolbar select2 fields are at least 150px wide

 * Including inline labels in toolbars

 * Prepended and appended inputs

 * Icon-only icons (they require font-size changes and a small offset)

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-checkout-restyle.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-checkout-restyle.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

oliverguenther committed:
Restyle navigation and instructons on show and empty

This commit provides:
* Adjustments on breadcrumbs according to styling
* Restored indent in repository table
* Disallow closing checkout instructions on empty page

The repository indent for opened folders was removed in the re-styling
of tables in
https://github.com/HDinger/openproject/commit/7a844aa9d5c2e18e3c737e83a3e85b348c87266b

oliverguenther committed:
Enable copy to clipboard through execCommand API

Uses execCommand API supported in some browsers to allow copy to
clipboard.
Displays a notification with the result of the operation

oliverguenther commented:
Thanks @basiszwo, there's been some inconsistent commits due to the intermittent changes to the visual. I've rebased them into a clear history.

oliverguenther committed:
Use new icon-compare in revision

oliverguenther commented:
Please note while reviewing. Firefox doesn't support the clipboard API until version 41. It will print an error notification if your Firefox is older than that.

NobodysNightmare commented:
![settings-button](https://cloud.githubusercontent.com/assets/453584/9657565/81e50932-5242-11e5-80d8-c9e3e8187072.png)

The behaviour of the settings button was unexpected to me. From other pages (WP list) I was used to the button opening a context menu, however: It immediately opened a new page.

I noticed that the subtle difference is the mssing dropdown arrow on your button. So it was probably somehow my fault. On the other hand: I can see a dropdown arrow in the visual.

oliverguenther commented:
The settings is meant to be a dropdown, however we neither have an available common style for use, nor is the dropdown specified. The agreed-upon alternative was this button for the time being.

NobodysNightmare commented on app/views/repositories/_breadcrumbs.html.erb:
```diff
@@ -27,30 +27,31 @@ See doc/COPYRIGHT.rdoc for more details.
 
-<%= link_to 'root', :action => 'show', :project_id => @project, :path => nil, :rev => @rev %>
+<%= link_to action: 'show', project_id: @project, path: nil, rev: @rev  do %>
+  <i class="icon-home2 repository-breadcrumbs--home"></i>
```
This one should carry a title and screenreaders should be able to read a description (e.g. `root directory`).

oliverguenther committed:
Work around IE bugs

Uses flex-basis and flex-grow to allow inputs to grow properly.

NobodysNightmare commented:
> Please note while reviewing. Firefox doesn't support the clipboard API until version 41. It will print an error notification if your Firefox is older than that.

Are you aware that this means, that I will receive the error message as soon as I click into the text field?
I understand the error message when I click on the button, but I would rather not display anything when copying by clicking into the text field failed.

To be clear: I like copying when clicking into the text field very much, I did not expect it, but was happy to see a message that told me it happened. But because it is quite unexpected for the copy to happen by clicking into the text field, it is even more unexpected to receive a red error message in that case.

This of course changes when I click on the button. My expectation is to have a copy of the text in my clipboard and I want to receive an error if that fails.

oliverguenther committed:
Accessibility and smaller corrections

Adds titles to:
* repository root breadcrumb
* copy to clipboard button

Corrections:
* Avoid copying to clipboard when clicking the input.
* Add hidden-for-sighted label for copy button.
* Remove notifications only when not in accessibility mode
* Avoid generic `-with-gap` button style
* Increase notification timeout

NobodysNightmare commented on app/views/repositories/_repository_header.html.erb:
```diff
@@ -0,0 +1,86 @@
+    <div class="toolbar-input--affix toolbar-input-group--affix -append"
+         copy-to-clipboard
+         clipboard-target="#repository-checkout-url">
+      <span class="icon-copy3"></span>
```
I'd also like some screen reader and title text here, to explain what this symbol means ("copy to clipboard")

oliverguenther replied:
added, thanks!

NobodysNightmare commented:
Did you notice that the select2 dropdowns do not match the width of the select2?

![image](https://cloud.githubusercontent.com/assets/453584/9658072/bc880f04-5246-11e5-97c2-6ab750050919.png)

oliverguenther commented:
I find that somehow inconsistent. I suggest we either remove the copy on click on the input itself, or leave the notification in there (it is readonly and clickable, so that doesn't come at such a surprise).

oliverguenther commented:
> Did you notice that the select2 dropdowns do not match the width of the select2?

I did, but I wondered that the actual select2 input field matches perfectly, and only the additional border seems off.

NobodysNightmare commented:
> it is readonly and clickable, so that doesn't come at such a surprise

I can only tell you that it surprised me... Maybe we should only keep the copy behaviour for the button, because it may be undesired in certain situations. Copying to the clipboard also means "deleting" the old contents of the clipboard.

> I did, but I wondered that the actual select2 input field matches perfectly, and only the additional border seems off.

That made me wonder, too... But I still think it is unintended ^^

oliverguenther commented:
> Maybe we should only keep the copy behaviour for the button, because it may be undesired in certain situations

With the specific button besides it, I can live with that. I'd suggest we only select & focus it upon clicking the input itself then?

NobodysNightmare commented:
> I'd suggest we only select & focus it upon clicking the input itself then?

Agreed.

Summarized:

**click on input**: select & focus text; no copy
**click on button**: copy text to clipboard; success/failure message; no select & focus

NobodysNightmare commented on app/assets/javascripts/repository_navigation.js:
```diff
@@ -68,6 +79,33 @@
+    Close checkout instructions
+    */
+    var checkout = $('#repository--checkout-instructions'),
+      toggle = $('#repository--checkout-instructions-toggle');
```
I think there are styling rules saying that `toggle` and `checkout` need to be aligned?

(from my point of view you could as well spam `var` keywords, but I think nobody else would like that ^^)

NobodysNightmare commented on app/assets/stylesheets/content/_buttons.sass:
```diff
@@ -63,6 +63,9 @@ a.button
     text-decoration: none
 
+  &.-with-gap
+    margin-top: 1rem
```
What makes me a little nervous is that this rule is more specific than it seems.

Gap? Of course on top... Gaps to the left would not make sense, would they?

But then I am irritated about buttons anyway. Apparently it makes sense that they only have margins to the right and at the bottom.

NobodysNightmare commented on app/assets/stylesheets/content/_notifications.sass:
```diff
@@ -49,6 +49,8 @@ $nm-color-warning-border: #b5b078
 $nm-color-warning-icon: #000
 $nm-color-warning-background: #f4f4aa
 
+$nm-color-info-icon: #000
```
I see that the warning icon also has `#000`, but I don't quite understand why those icons have a different color than the text. (because the difference is really marginal)

oliverguenther replied:
It's proably undesired. I'll change it to #333

NobodysNightmare commented on app/assets/stylesheets/layout/_toolbar.sass:
```diff
@@ -36,6 +36,11 @@
+@mixin toolbar-element-sizing
+  margin: 0.625rem 0 0
+  height: 40px
+  line-height: 22px
```
What is the purpose of this one?
I tend to get nervous when seeing changes to the line-height, since I only saw bad things happen when the line-height was touched.

On the bright side: I can't see any difference when enabling/disabling this rule (that means no bad things I could observe).

edit: jsut realized that you copied this from somewhere else... but maybe you understand it ^^

oliverguenther replied:
I think that this was meant to vertical center the icon spans in some of the buttons. I'll re-evaluate whether this is still required (but yes, I simply moved it into a mixin)

oliverguenther replied:
If you omit the line-height, smaller icons not contained immediately in a toolbar item are no longer centered (this still applies to the clipboard symbol). For some other icons were this was a problem before, this no longer applies to due switching to flex.

NobodysNightmare replied:
Thanks for your research efforts!

NobodysNightmare commented on app/assets/stylesheets/open_project_global/_variables.sass:
```diff
@@ -189,3 +194,5 @@ $user-avatar-mini-border-radius:                50px !default
 
 $select-element-padding:                        3px, 24px, 3px, 3px
+
+$repository-breadcrumbs--color:                 #78818A
```
First: not sure if this belongs into the global variables, but be it...

What I do not understand (but sadly this is what the visual says) is why we need to introduce a new color here. Now we have three colors:

![three-colors](https://cloud.githubusercontent.com/assets/453584/9658898/2a06d9f6-524d-11e5-8d18-8355a7e97e08.PNG)
- root
- clickable/linked directory (ancestor)
- current directory/file (no link)

when we would only need two colors:

![two-colors](https://cloud.githubusercontent.com/assets/453584/9658944/6e3824fe-524d-11e5-804f-0c641d07f8e5.PNG)
- clickable
- non-clickable

The new color does exist nowhere else (_inconsistency_) and I am asking myself if it is the same color on magenta themed OpenProject instances.

Bonus: The underline color is currently wrong:

![wrong underline](https://cloud.githubusercontent.com/assets/453584/9659007/d9762176-524d-11e5-8590-22b4ad4ab101.png)

oliverguenther replied:
Agreed, I'm ruling this should be the same link color. But note that there is still the small separator icon that had the same color. So in fact, even with your suggestion we have three colors (but I'm setting that to gray-dark, which is very similar).

NobodysNightmare commented on app/views/repositories/_checkout_instructions.html.erb:
```diff
@@ -1,30 +1,33 @@
-          <% end %>
+<div id="repository--checkout-instructions" class="notification-box -info">
+  <% unless embedded %>
+    <a href="javscript:" class="notification-box--close">&times;</a>
```
Shouldn't you be using `icon-close2` or a similar icon here?

Maybe this is a more general issue/discussion point, out of the scope of this PR. I just stumbled upon that.

NobodysNightmare replied:
Okay, I just tested this using chromevox, which reads "Multiplication" when tabbing onto that link.

oliverguenther replied:
It is consistent used in the styleguide, but I agree. We should make this consistent in a subsequent PR.

NobodysNightmare commented on app/views/repositories/_checkout_instructions.html.erb:
```diff
@@ -1,30 +1,33 @@
-              <strong><%= l("repositories.checkout.access.#{instructions.permission}") %></strong>
-            </p>
-          <% end %>
+<div id="repository--checkout-instructions" class="notification-box -info">
```
I just realized that this file has no copyright header :scream:

basiszwo replied:
as for copy to clipboard: can't we use a solution that works in all browsers?

http://zeroclipboard.org/ this is what e.g. github uses (afaik)

oliverguenther replied:
Yes, but with the recent announcement of chrome and firefox, this method will no longer work given the forced switch to 'click to play' mechanics of flash.

This does work in the next stable releases of Firefox, the current stables of IE and Chrome. Seems plenty to me and orders of magnitude better than having a flash dependency.

NobodysNightmare replied:
I just remember an old article by github saying that they use flash for their clipboard actions, since this is the only cross browser way to do it...

If there is a new API, that will "soon" be supported by all browsers, I'd say this is desirable, even if not perfect today.

oliverguenther replied:
Reference: https://hacks.mozilla.org/2015/09/flash-free-clipboard-for-the-web/

> Firefox 41 ... release in mid-September, it is becoming available to JavaScript within user-action initiated callbacks.
> 
> Google Chrome and Internet Explorer both also support this API. Chrome uses the same restriction as Firefox (that it must be run in a user-initiated callback). Internet Explorer allows it to be called at any time, except it first prompts the user with a dialog, asking for permission to access the clipboard.

I'm assuming other browsers will probably follow this move.

NobodysNightmare commented on app/views/repositories/_repository_header.html.erb:
```diff
@@ -0,0 +1,88 @@
+           readonly
+           copy-to-clipboard>
+    <button class="toolbar-input--affix toolbar-input-group--affix -append"
+            title="<%= l(:label_copy_to_clipboard) %>"
```
Not sure if it is enough to set the title only and not including a `<span class="hidden-for-sighted">some text</span>`.

ChromeVox seems to read it out correctly, but I can't test Jikes and don't know our exact guidelines here...

NobodysNightmare commented:
I was browsing our [Accessibility Checklist](https://www.openproject.org/open-source/development/accessibility-checklist/) looking for an indication whether the `title` is enough context. I could not find anything on that, however I found another interesting paragraph:

> - If a page or application has a time limit, the user is given options to turn off, adjust, or extend that time limit.
> - The content and functionality has no time limits or constraints.

I believe the one second time limit on the copy success message and the 3 second limit on the error message could violate this constraint.

At the very least the time limits should be disabled in accessibility mode, maybe even always...

NobodysNightmare commented on app/views/repositories/annotate.html.erb:
```diff
@@ -26,9 +26,15 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+<%= call_hook(:view_repositories_show_contextual, { repository: @repository, project: @project }) %>
+<%= render partial: 'repository_header', locals: { empty: false } %>
+
+<%# Breadcrumbs %>
```
obvious comment is obvious?

oliverguenther commented:
Disabling the timeout for all users is an incredible nuissance and I won't have that. I will disable them when accessibility mode is enabled, though.

NobodysNightmare commented on config/locales/js-en.yml:
```diff
@@ -32,6 +32,11 @@ en:
     close_popup_title: "Close popup"
+
+    clipboard:
+      browser_error: "Your browser doesn't seem to support copying to clipboard. Please copy the selected text manually."
```
Minor:

> Your browser doesn't ~~seem to~~ support copying to clipboard.

I personally also tend to weaken my messages ("I am not sure, but it looks like..."), but thinking about it: No! that stupid browser does not work as our application wants it to work, so we can tell that.

oliverguenther commented:
I've disabled the removal timeout on accessbility mode, however it seems that `ConfigurationService.accessibilityModeEnabled()` doesn't evaluate to true when I alter the settings under my account.

NobodysNightmare commented:
> Disabling the timeout for all users is an incredible nuissance and I won't have that. I will disable them when accessibility mode is enabled, though.

I'd nevertheless think about longer timeouts. One second is rather short:
1. move your eyes over from the button to the message
2. really read it (as if you wouldn't know whats written there)
3. assume that another language might not be as concise and requires more reading time (or the reader is just slow)

edit: I just used @Azure7111 as test subject, he agrees that one second is too short ^^

oliverguenther commented:
> I'd nevertheless think about longer timeouts. One second is rather short.

Then click the button again? :grimacing: But yeah, we can set it higher. 3000 ? or make sure and go over 9000 ?

NobodysNightmare commented:
**_WHAT?!?!? NINE THOUSAND?!?!?!**_

...

Sorry, I couldn't resist :smile: 

I would also agree with 5000, but I actually wouldn't even care if it was infinite ^^

oliverguenther commented:
5000 horsies it is, but not one more! :horse:

oliverguenther commented:
@NobodysNightmare I think I've addressed all remarks you mentioned. To summarize:
- select2 dropdown: fixed border
- click on input == selection, click on clipboard icon == copy
- clipboard button: title and hidden-for-sighted  span
- Moved the `-with-gap` edgecase button into specific scm style
- Increased notification timeout in ms to ~~over 9000~~ 5000

NobodysNightmare commented:
Sorry. I found one more issue in IE 11.

### IE 11

![ie11](https://cloud.githubusercontent.com/assets/453584/9662506/fb54adbe-525f-11e5-87a0-803887946344.PNG)

The textbox containing the URL is too small.

edit: looking harder at both screenshots, I can now also see that the Icons inside the buttons are misaligned in IE...

### Expected (Chrome)

![expected](https://cloud.githubusercontent.com/assets/453584/9662505/fb52c544-525f-11e5-9c86-54fc21eb3fed.PNG)

oliverguenther commented:
I couldn't figure this out but provided a workaround (minimal width) as the flexbox width doesn't seem to be determined properly in IE 11, which is why you're seeing the minimal width on 11. :confused: 

Maybe @HDinger has some input on that?

NobodysNightmare commented:
I've extracted tickets for the remaining issues:
- [close with multiplication sign](https://community.openproject.org/work_packages/21407)
- [IE11 textbox too small](https://community.openproject.org/work_packages/21408)

NobodysNightmare commented:
Apart from the extracted issues this should be good to merge! :horse: 

Lets wait for travis and then merge it! :horse_racing:
